packages/aws-cdk-lib/aws-cloudwatch/lib/variable.ts

```diff
@@ -0,0 +1,195 @@
+   */
+  readonly type: VariableType;
+
+  /**
+   * The way the variable value is selected
+   */
+  readonly inputType: VariableInputType;
+
+  /**
+   * Pattern or property value to replace
+   */
+  readonly value: string;
+
+  /**
+   * Unique id
+   */
+  readonly id: string;
+
+  /**
+   * Optional label in the toolbar
+   *
+   * @default - the variable's value
+   */
+  readonly label?: string;
+
+  /**
+   * Optional default value
+   *
+   * @default - no default value is set
+   */
+  readonly defaultValue?: any;
+
+  /**
+   * Whether the variable is visible
+   *
+   * @default - true
+   */
+  readonly visible?: boolean;
+}
+
+export abstract class DashboardVariable implements IVariable {
+  private readonly baseOptions: DashboardVariableOptions;
+
+  protected constructor(options: DashboardVariableOptions) {
+    this.baseOptions = options;
+  }
+
+  toJson(): any {
+    return {
+      [this.baseOptions.type]: this.baseOptions.value,
+      type: this.baseOptions.type,
+      inputType: this.baseOptions.inputType,
+      id: this.baseOptions.id,
+      defaultValue: this.baseOptions.defaultValue,
+      visible: this.baseOptions.visible,
+      label: this.baseOptions.label,
+    };
+  }
+}
+
+export interface VariableValue {
+  /**
+   * Optional label for the selected item
+   *
+   * @default - the variable's value
+   */
+  readonly label?: string;
+
+  /**
+   * Value of the selected item
+   */
+  readonly value: string;
+}
+
+/**
+ * Options for {@link ValueDashboardVariable}
+ */
+export interface ValueDashboardVariableOptions extends DashboardVariableOptions {
+  /**
+   * List of custom values for the variable.
+   * It is required for variables of types {@link VariableInputType.RADIO} and {@link VariableInputType.SELECT}
+   *
+   * @default - no values
+   */
+  readonly values?: VariableValue[];
+}
+
+/**
+ * A dashboard variable supporting all {@link VariableInputType}.
+ */
+export class ValueDashboardVariable extends DashboardVariable {
+  private readonly options: ValueDashboardVariableOptions;
+
+  constructor(options: ValueDashboardVariableOptions) {
+    super(options);
+    if (options.inputType != VariableInputType.INPUT && (options.values || []).length == 0) {
+      throw new Error(`Variable with input type ${options.inputType} requires values to be provided.`);
+    }
+    this.options = options;
+  }
+
+  toJson(): any {
+    const base = super.toJson();
+    return {
+      ...base,
+      values: this.options.values ? this.options.values.map(value => ({ label: value.label, value: value.value })) : undefined,
+    };
+  }
+}
+
+/**
+ * Options for {@link SearchDashboardVariable}
+ */
+export interface SearchDashboardVariableOptions extends DashboardVariableOptions {
+  /**
+   * A search expression that specifies a namespace, dimension name and a metric name.
+   *
+   * For example `{AWS/EC2,InstanceId} MetricName=\"CPUUtilization\"`
+   */
+  readonly searchExpression: string;
```

Comment:
Have you given any thought to strongly typing this expression? How unique can this string be? The examples in documentation all reference a namespace, dimension, and metricName in some format. Maybe with stronger types we can build this expression out for the user and not have to have them worry about typos.

I'm thinking something like:

```ts
class searchExpression {
  public static FROM_COMPONENTS(
    namespace: string,
    dimension: string,
    metricName: string,
  ) {
    return new searchExpression(`{${namespace},${dimension}} MetricName=\"${metricName\"`);
  }

  // users can still supply raw expressions if they want
  public constructor(public readonly expression: string) {}
}
```

WDYT?

Reply:
So, I'd be a bit hesitant here, without further investigation, abt how metrics with multiple dimensions work - if they're supported or not.
If you think it's needed, I can give that a try and see if that's supported, in which case the search expression would need to allow for number of dimensions >= 1

Reply:
Great! I think if possible, a class would be a better customer experience than a raw string.

Reply:
I blame you for the more code - now it needs more reviewing 😆 

https://github.com/aws/aws-cdk/pull/26285/commits/d8cc14a9494c9b907248b2437124a7b8557b512a

I ended up refactoring how search values are modeled!

Reply:
In fact, with this new change, I'm now wondering if we really need a separate `ValueDashboardVariable` and `SearchDashboardVariable`. The difference can really be just about the `values` prop. Something like `Values.fromSearch` and `Values.fromStatic` (or any better name).

Worth pursuing?

Reply:
I went ahead and did that. I believe it looks better.
Let me know what you think.
I also believe docs/README/error messages might need a good look to ensure they make sense for users.